### PR TITLE
Star images

### PR DESCRIPTION
**Pull request by maryhipp**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Community Node Submission

## Have you discussed this change with the InvokeAI team?
- [x] Yes
- [ ] No, because:

      
## Have you updated all relevant documentation?
- [ ] Yes
- [ ] No

## Description
Add the ability to "star" and "unstar" images, either one-by-one or via multiselect.
Changes the delete icon functionality - removes it as an optional setting, and adds interaction to have it show up on shift+hover.

## Related Tickets & Documents

- Related Issue #
- Closes #

## Added/updated tests?

- [ ] Yes
- [ ] No : _please replace this line with details on why tests
      have not been included_

### CODE REVIEW

File: invokeai/frontend/web/src/features/gallery/components/ImageGrid/GalleryImage.tsx [hunk 1]
```diff
@@ -80,15 +91,17 @@ const GalleryImage = (props: HoverableImageProps) => {
...
+          onClickReset={togglePinnedState}
           imageSx={{ w: 'full', h: 'full' }}
           isDropDisabled={true}
           isUploadDisabled={true}
           thumbnail={true}
           withHoverOverlay
-          resetIcon={<FaTrash />}
-          resetTooltip="Delete image"
-          withResetIcon={shouldShowDeleteButton} // removed bc it's too easy to accidentally delete images
+          resetIcon={
```
Comment (maryhipp) [inline invokeai/frontend/web/src/features/gallery/components/ImageGrid/GalleryImage.tsx:100]: @psychedelicious any thoughts on this interaction? Would need to remove the option to have delete icon here

File: invokeai/frontend/web/src/services/api/endpoints/images.ts [hunk 2]
```diff
@@ -387,6 +387,116 @@ export const imagesApi = api.injectEndpoints({
...
+
+        const isCacheFullyPopulated =
+          currentCache.data && currentCache.data.ids.length >= (total ?? 0);
+
+        const isInDateRange = getIsImageInDateRange(
+          currentCache.data,
+          updatedImage
+        );
+
+        // should we remove images from cache if _not_ in date range? ie you are showing 100 of 101 pinned images and you unpin one. technically it should disappear from list.
```
Comment (maryhipp) [inline invokeai/frontend/web/src/services/api/endpoints/images.ts:476]: @psychedelicious don't mean to trigger you on more gallery cache stuff but lmk what you think

File: invokeai/frontend/web/src/features/gallery/components/ImageGrid/GalleryImage.tsx [hunk 3]
```diff
@@ -80,16 +107,34 @@ const GalleryImage = (props: HoverableImageProps) => {
...
+          onMouseOut={() => setIsHovered(false)}
+        >
+          <>
+            <IAIDndImageIcon
+              onClick={toggleStarredState}
+              icon={starIcon}
+              tooltip={imageDTO.starred ? 'Unstar' : 'Star'}
+            />
+
+            {isHovered && shouldShowDeleteButton && (
```
Comment (psychedelicious) [inline invokeai/frontend/web/src/features/gallery/components/ImageGrid/GalleryImage.tsx:125]: We track `shift` globally, it's in `state.hotkeys.shift`

**Review by psychedelicious — APPROVED** [review]

**Comment by psychedelicious** [comment]
I made a few small changes/fixes and rebased on main:

[feat(ui): show Star All if selection is a mix of starred and unstarred](https://github.com/invoke-ai/InvokeAI/pull/4246/commits/1367136d1b7e052c411d3817571790e9c187e6cf)

[fix: add response model for star/unstar routes](https://github.com/invoke-ai/InvokeAI/pull/4246/commits/b46d0b60c7d8f7f6d72593bf873c7c0ec335cb1a)


- also implement pessimistic updates for starring, only changing the images that were successfully updated by backend
- some autoformat changes crept in

[feat(ui): use imageDTOs instead of images in starring queries](https://github.com/invoke-ai/InvokeAI/pull/4246/commits/edfa5a7892a019fa87111cdbc584223231992090)

[fix(ui): restore reset button in node image component](https://github.com/invoke-ai/InvokeAI/pull/4246/commits/468a464a148f808da68d7386812f0a739576182c)

[feat(ui): remove shouldShowDeleteButton](https://github.com/invoke-ai/InvokeAI/pull/4246/commits/7fe482cd42e1139efe62477ae4e5bdd7ede335ac)


- remove this state entirely
- use `state.hotkeys.shift` directly to hide and show the icon on gallery
- also formatting

**Pull request merged**